src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultCustomCredentialPolicy.java

```diff
@@ -0,0 +1,117 @@
+import com.azure.core.util.CoreUtils;
+
+import java.text.MessageFormat;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import reactor.core.publisher.Mono;
+
+/**
+ * A policy that authenticates requests with Azure Key Vault service.
+ */
+class KeyVaultCustomCredentialPolicy implements HttpPipelinePolicy {
+    private static final String WWW_AUTHENTICATE = "WWW-Authenticate";
+    private static final String BEARER_TOKEN_PREFIX = "Bearer ";
+    private static final String AUTHORIZATION = "Authorization";
+    private final ScopeTokenCache cache;
+    private final KeyVaultTokenCredential keyVaultTokenCredential;
+
+    /**
+     * Creates KeyVaultCustomCredentialPolicy.
+     *
+     * @param credential
+     *        the token credential to authenticate the request
+     * @throws SQLServerException
+     */
+    KeyVaultCustomCredentialPolicy(KeyVaultTokenCredential credential) throws SQLServerException {
+        if (null == credential) {
+            MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_NullValue"));
+            Object[] msgArgs1 = {"Credential"};
+            throw new SQLServerException(form.format(msgArgs1), null);
+        }
+
+        this.cache = new ScopeTokenCache(credential::getToken);
+        this.keyVaultTokenCredential = credential;
+    }
+
+    /**
+     * Adds the required header to authenticate a request to Azure Key Vault service.
+     *
+     * @param context
+     *        The request context
+     * @param next
+     *        The next HTTP pipeline policy to process the {@code context's} request after this policy completes.
+     * @return A {@link Mono} representing the HTTP response that will arrive asynchronously.
+     */
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+        if ("http".equals(context.getHttpRequest().getUrl().getProtocol())) {
+            return Mono.error(new RuntimeException(SQLServerException.getErrString("R_TokenRequireUrl")));
+        }
+
+        return next.clone().process()
+                // Ignore body
+                .doOnNext(HttpResponse::close).map(res -> res.getHeaderValue(WWW_AUTHENTICATE))
+                .map(header -> extractChallenge(header, BEARER_TOKEN_PREFIX)).flatMap(map -> {
+                    keyVaultTokenCredential.setAuthorization(map.get("authorization"));
+                    keyVaultTokenCredential.setResource(map.get("resource"));
+                    keyVaultTokenCredential.setScope(map.get("scope"));
+                    cache.setRequest(new TokenRequestContext().addScopes(map.get("resource") + "/.default"));
+                    return cache.getToken();
+                }).flatMap(token -> {
+                    context.getHttpRequest().setHeader(AUTHORIZATION, BEARER_TOKEN_PREFIX + token.getToken());
+                    return next.process();
+                });
```

Comment:
This is very interesting syntax but can we either link a source for a spec detailing why we want these specific fields or add some comments to explain. The string field names should also be changed to constants incase it changes in the future.